File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-swift/Sources/MailSdk/mailProviderExtensionCatalog.swift
public struct MailProviderExtensionCatalogEntry {
    public let extensionKey: String
    public let providerKey: String
    public let displayName: String
    public let surface: String
    public let access: String
    public let status: String
}

public enum MailProviderExtensionCatalog {
    public static let recognizedSurfaces: [String] = [
        "control-plane",
        "runtime-bridge",
        "cross-surface",
    ]

    public static let recognizedAccessModes: [String] = [
        "unwrap-only",
        "extension-object",
    ]

    public static let recognizedStatuses: [String] = [
        "reference-baseline",
        "reserved",
    ]

    public static let entries: [MailProviderExtensionCatalogEntry] = [
        .init(extensionKey: "smtp.transport", providerKey: "smtp", displayName: "SMTP Transport", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "imap.sync", providerKey: "imap", displayName: "IMAP Sync", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
    ]

public static func getMailProviderExtensionCatalog() -> [MailProviderExtensionCatalogEntry] {
        entries
    }

    public static func getMailProviderExtensionDescriptor(_ extensionKey: String) -> MailProviderExtensionCatalogEntry? {
        entries.first { $0.extensionKey == extensionKey }
    }

    public static func getMailProviderExtensionsForProvider(_ providerKey: String) -> [MailProviderExtensionCatalogEntry] {
        entries.filter { $0.providerKey == providerKey }
    }

    public static func getMailProviderExtensions(_ extensionKeys: [String]) -> [MailProviderExtensionCatalogEntry] {
        var resolved: [MailProviderExtensionCatalogEntry] = []
        for extensionKey in extensionKeys {
            if let entry = getMailProviderExtensionDescriptor(extensionKey) {
                resolved.append(entry)
            }
        }

        return resolved
    }

    public static func hasMailProviderExtension(_ extensionKeys: [String], _ extensionKey: String) -> Bool {
        extensionKeys.contains(extensionKey) && getMailProviderExtensionDescriptor(extensionKey) != nil
    }

}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-kotlin/src/main/kotlin/com/sdkwork/Mail/metadata/mailProviderExtensionCatalog.kt
package com.sdkwork.Mail.metadata

data class MailProviderExtensionCatalogEntry(
    val extensionKey: String,
    val providerKey: String,
    val displayName: String,
    val surface: String,
    val access: String,
    val status: String,
)

object MailProviderExtensionCatalog {
    val recognizedSurfaces: List<String> = listOf(
        "control-plane",
        "runtime-bridge",
        "cross-surface",
    )

    val recognizedAccessModes: List<String> = listOf(
        "unwrap-only",
        "extension-object",
    )

    val recognizedStatuses: List<String> = listOf(
        "reference-baseline",
        "reserved",
    )

    val entries: List<MailProviderExtensionCatalogEntry> = listOf(
        MailProviderExtensionCatalogEntry("smtp.transport", "smtp", "SMTP Transport", "runtime-bridge", "unwrap-only", "reserved"),
        MailProviderExtensionCatalogEntry("imap.sync", "imap", "IMAP Sync", "runtime-bridge", "unwrap-only", "reserved"),
    )

fun getMailProviderExtensionCatalog(): List<MailProviderExtensionCatalogEntry> = entries

    fun getMailProviderExtensionDescriptor(extensionKey: String): MailProviderExtensionCatalogEntry? =
        entries.firstOrNull { it.extensionKey == extensionKey }

    fun getMailProviderExtensionsForProvider(providerKey: String): List<MailProviderExtensionCatalogEntry> =
        entries.filter { it.providerKey == providerKey }

    fun getMailProviderExtensions(extensionKeys: List<String>): List<MailProviderExtensionCatalogEntry> =
        extensionKeys.mapNotNull(::getMailProviderExtensionDescriptor)

    fun hasMailProviderExtension(extensionKeys: List<String>, extensionKey: String): Boolean =
        extensionKeys.contains(extensionKey) && getMailProviderExtensionDescriptor(extensionKey) != null

}
